### PR TITLE
Logger fix crash after %s parsing and memory leak

### PR DESCRIPTION
**Pull request by ktrzcinx**
output after changes:

``` 
[122156200059.062500] (      171.250000) c0 pipe                   src/audio/pipeline.c:78   pipeline new pipe_id 6 period 1000 priority 0
error: String printing is not supported
[122156200249.375000] (      190.312500) c0 component             src/audio/component.c:130  comp new (null) type 1 id 7.34
[122156200269.895844] (       20.520834) c0 dma                           src/lib/dma.c:126  dma_get() ID 5 sref = 5 busy channels 0
error: String printing is not supported
[122156200428.020844] (      158.125000) c0 component             src/audio/component.c:130  comp new (null) type 5 id 7.35
[122156200447.187500] (       19.166666) c0 pga          7.35  ../audio/volume/volume.c:375  vol->initial_ramp = 250, vol->ramp = 0, vol->min_value = 0, vol->max_value = 65536
[122156200573.281250] (      126.093750) c0 buffer                   src/audio/buffer.c:79   buffer new size 0x300 id 7.36 flags 0x0
```

### CODE REVIEW

File: tools/logger/convert.c [hunk 1]
```diff
@@ -198,6 +198,12 @@ static void process_params(struct proc_ldc_entry *pe,
 		if (p[1] == '%') {
 			/* Skip "%%" */
 			p += 2;
+		} else if (p[1] == 's') {
+			/* check for string printing, because it leads to logger crash */
+			log_err("String printing is not supported\n");
```
Comment (lgirdwood) [inline tools/logger/convert.c:203]: Is there anyway we can print the line and file where the trace is attempting a string (to help the developer fix).

File: tools/logger/convert.c [hunk 2]
```diff
@@ -849,19 +849,25 @@ int convert(struct convert_config *config)
 		      config->ldc_fd);
 	if (!count) {
 		log_err("failed to read uuid section data.\n");
-		return -ferror(config->ldc_fd);
+		ret = -ferror(config->ldc_fd);
```
Comment (lgirdwood) [inline tools/logger/convert.c:852]: Sorry, not following the commit message. What is leaking and how are we fixing it.

File: tools/logger/convert.c (hunk 1 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:203]: Full message (including trace file:line location in source code) is printed.

File: tools/logger/convert.c (hunk 2 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:852]: Changed commit message

**Review by lgirdwood — APPROVED** [review]

File: tools/logger/convert.c [hunk 3]
```diff
@@ -131,14 +131,48 @@ static const char *format_uid(uint32_t uid_ptr, int use_colors, bool be, bool up
 	return str;
 }
 
+/* fmt should point '%pUx`, return length of UUID format string or zero */
+static int get_uuid_format(const char *fmt, const char *fmt_end, int *be, int *upper)
```
Comment (paulstelian97) [inline tools/logger/convert.c:135]: I'd like snprintf semantics: You write as much as you can and return the length of the would-be UUID string. The caller is responsible for reallocating if necessary.

File: tools/logger/convert.c [hunk 4]
```diff
@@ -198,6 +198,12 @@ static void process_params(struct proc_ldc_entry *pe,
 		if (p[1] == '%') {
 			/* Skip "%%" */
 			p += 2;
+		} else if (p[1] == 's') {
+			/* check for string printing, because it leads to logger crash */
+			log_err("String printing is not supported\n");
+			pe->params[i] = 0;
```
Comment (paulstelian97) [inline tools/logger/convert.c:204]: I'd actually like to see something like "<String @ 0x924046F9>" or something, to distinguish between compile time strings and runtime strings. Not mandatory but nice to have.

Comment (ktrzcinx) [reply tools/logger/convert.c]: Added

File: tools/logger/convert.c (hunk 3 above)
Comment (ktrzcinx) [reply tools/logger/convert.c]: changed to `asprintf_uuid`.

File: tools/logger/convert.c [hunk 5]
```diff
@@ -131,14 +131,53 @@ static const char *format_uid(uint32_t uid_ptr, int use_colors, bool be, bool up
...
+/* fmt should point '%pUx`, return pointer to UUID string or zero */
+static const char *asprintf_uuid(const char *fmt, uint32_t uuid_key, bool use_colors, int *fmt_len)
+{
+	const char *fmt_end = fmt + strlen(fmt);
+	int be, upper;
+	int len = 4; /* assure full formating, with x */
+
+	if (fmt + 2 >= fmt_end || fmt[1] != 'p' || fmt[2] != 'U') {
+		*fmt_len = 0;
+		return 0;
```
Comment (paulstelian97) [inline tools/logger/convert.c:143]: Shouldn't you return NULL instead of 0 here? The return type is `const char *`.

File: tools/logger/convert.c [hunk 6]
```diff
@@ -131,14 +131,53 @@ static const char *format_uid(uint32_t uid_ptr, int use_colors, bool be, bool up
...
+		upper = 1;
+		break;
+	default:
+		be = 0;
+		upper = 0;
+		--len;
+		break;
+	}
+	*fmt_len = len;
+	return format_uid(uuid_key, use_colors, be, upper);
```
Comment (paulstelian97) [inline tools/logger/convert.c:170]: Does format_uid allocate a new string on its own, to obey the asprintf semantics?

File: tools/logger/convert.c [hunk 7]
```diff
@@ -200,6 +200,13 @@ static void process_params(struct proc_ldc_entry *pe,
 		if (p[1] == '%') {
 			/* Skip "%%" */
 			p += 2;
+		} else if (p[1] == 's') {
+			/* check for string printing, because it leads to logger crash */
+			log_err("String printing is not supported\n");
+			pe->params[i] = (uintptr_t)asprintf("<String @ 0x%08x>", e->params[i]);
```
Comment (paulstelian97) [inline tools/logger/convert.c:206]: Acceptable memory leak?

Comment (ktrzcinx) [reply tools/logger/convert.c:206]: subst_mask is set, so there shouldn't be memory leak

File: tools/logger/convert.c (hunk 6 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:170]: yes, format_uid use asprintf internally.

File: tools/logger/convert.c (hunk 7 above)
Comment (paulstelian97) [reply tools/logger/convert.c:206]: Oh, so you use subst_mask to clean this up?

Comment (ktrzcinx) [reply tools/logger/convert.c:206]: It has been implemented before,
```
static void free_proc_ldc_entry(struct proc_ldc_entry *pe)
{
	int i;

	for (i = 0; i < TRACE_MAX_PARAMS_COUNT; i++) {
		if (pe->subst_mask & (1 << i))
			free((void *)pe->params[i]);
		pe->params[i] = 0;
	}
}
```

File: tools/logger/convert.c (hunk 5 above)
Comment (ktrzcinx) [reply tools/logger/convert.c]: replaced

**Review by paulstelian97 — APPROVED** [review]

**Review by mmaka1 — APPROVED** [review]

**Review by akloniex — APPROVED** [review]

**Pull request merged**